docs/auto-generating-selectors.md

```diff
@@ -11,27 +11,28 @@ However, writing these could be tedious, but you can auto-generate them
 ## create the following function: `createSelectors`
 
 ```typescript
-import { State, StoreApi, UseBoundStore } from 'zustand'
+import { State, StoreApi, UseBoundStore } from "zustand";
 
-interface Selectors<StoreType> {
-  use: {
-    [key in keyof StoreType]: () => StoreType[key]
-  }
-}
+type WithSelectors<S> = S extends { getState: () => infer T }
+  ? S & { use: { [K in keyof T]: () => T[K] } }
+  : never
 
-export default function createSelectors<StoreType extends State>(
-  store: UseBoundStore<StoreType, StoreApi<StoreType>>
-) {
-  // Casting to any to allow adding a new property
-  ;(store as any).use = {}
+type CreateSelectors =
+  <S extends UseBoundStore<StoreApi<State>>>
+  (s: S) =>
+    WithSelectors<S>
 
-  Object.keys(store.getState()).forEach((key) => {
-    const selector = (state: StoreType) => state[key as keyof StoreType]
-    ;(store as any).use[key] = () => store(selector)
-  })
+const createSelectors: CreateSelectors = _store => {
+  type S = typeof _store
+  type T = S extends { getState: () => infer T } ? T : never
+
+  let store = _store as WithSelectors<S>
```

Comment:
Let's make it more compact

```suggestion
let store = _store as WithSelector<typeof _store>
```

Reply:
You also might want to merge the type declaration and implementation in case that's the convention followed elsewhere...

```ts
const createSelector = <S extends UseBoundStore<StoreApi<State>>>(_store: S) =>
...
```

It's just that I was commenting on the issue and my style is to separate types and implementations.